src/widget/scroll.rs

```diff
@@ -239,30 +271,104 @@ impl<T: Data, W: Widget<T>> Widget<T> for Scroll<T, W> {
     fn event(&mut self, event: &Event, ctx: &mut EventCtx, data: &mut T, env: &Env) {
         let size = ctx.base_state.size();
         let viewport = Rect::from_origin_size(Point::ORIGIN, size);
-        let child_event = event.transform_scroll(self.scroll_offset, viewport);
-        if let Some(child_event) = child_event {
-            self.child.event(&child_event, ctx, data, env)
-        };
-
-        match event {
-            // Show the scrollbars any time our size changes
-            Event::Size(_) => self.reset_scrollbar_fade(ctx),
-            // The scroll bars will fade immediately if there's some other widget requesting animation.
-            // Guard by the timer id being invalid.
-            Event::AnimFrame(interval) if self.scroll_bars.timer_id == TimerToken::INVALID => {
-                // Animate scroll bars opacity
-                let diff = 2.0 * (*interval as f64) * 1e-9;
-                self.scroll_bars.opacity -= diff;
-                if self.scroll_bars.opacity > 0.0 {
-                    ctx.request_anim_frame();
+
+        if match event {
+            Event::MouseMoved(_) | Event::MouseUp(_) => {
+                self.scroll_bars.vertical_held || self.scroll_bars.horizontal_held
+            }
+            _ => false,
```

Comment:
one of the powers of rust match statements is that you can include conditionals in each arm; arms are evaluated in order, and the first matching arm is followed. So you can match the same item with different conditions, which should let you clean this up quite a bit, like:

```rust
match event {
    Event::MouseMoved(_) if self.scroll_bars.vertical_held || self.scroll_bars.horizontal_held => ...,
    Event::MouseUp(_) => if self.scroll_bars.vertical_held || self.scroll_bars.horizontal_held => ...,
    Event::MouseMoved(_) => ... // only if we didn't match above,
    Event::MouseUp(_) => ... // only if we didn't match above,
    _ => (), // all other events
}

Reply:
Noice :ok_hand:

Reply:
rustfmt pls
![image](https://user-images.githubusercontent.com/30666851/68211065-d08afd80-ffa4-11e9-895f-fd9ecd648550.png)

Reply:
I think in this specific spot the original was better, I'll try to find any other match cases I added where it could be applied though

Reply:
Hmm. My thinking with that initial comment was that you could replace _all_ instances of `if match` with a single match statement that had conditionals on the arms?

This isn't strictly necessary, but I think that `if match()` is hard to follow, in terms of the control flow, because you have to jump from looking at the match statement down to the `{` where the conditional statement is, often jumping back up to look at various arms of the match to see which are evaluating true and false; if this is a single `match` statement you can basically read it top to bottom, following immediately from a match to the branch that will be taken.


... _some time passes_ ...

okay, I wanted to actually see if my ideas made sense, since this does seem like tricky control flow. Here is my best attempt at implementing the `event` fn, which i _think_ has identical behaviour to your implementation; hopefully this can be helpful at least as an illustration. I personally think it's slightly clearer to follow, but not hugely so or anything; the logic here is hard to capture.

```rust
fn event(&mut self, event: &Event, ctx: &mut EventCtx, data: &mut T, env: &Env) {
    let size = ctx.base_state.size();
    let viewport = Rect::from_origin_size(Point::ORIGIN, size);
    
    // first check if this is a mouse even related to the scroll bars,
    // which we will swallow.
    let is_handled = match event {
        Event::MouseMoved(event) if self.scroll_bars.held.is_held() => {
            self.reset_scrollbar_fade(ctx);
            match self.scroll_bars.held {
                BarHeldState::VerticalHeld(offset) => {
                    let scale_y = viewport.height() / self.child_size.height;
                    let bounds = self.calc_vertical_bar_bounds(&viewport);
                    let mouse_y = event.pos.y + self.scroll_offset.y;
                    let delta = mouse_y - bounds.y0 - offset;
                    self.scroll(Vec2::new(0f64, (delta / scale_y).ceil()), size);
                    true
                }
                BarHeldState::HorizontalHeld(offset) => {
                    let scale_x = viewport.width() / self.child_size.width;
                    let bounds = self.calc_horizontal_bar_bounds(&viewport);
                    let mouse_x = event.pos.x + self.scroll_offset.x;
                    let delta = mouse_x - bounds.x0 - offset;
                    self.scroll(Vec2::new((delta / scale_x).ceil(), 0f64), size);
                    true
                }
                _ => unreachable!(),
            }
        }
        Event::MouseUp(_) if self.scroll_bars.held.is_held() => {
            self.scroll_bars.held = BarHeldState::None;
            true
        }
        Event::MouseMoved(event) => {
            let transformed_pos = event.pos + self.scroll_offset;
            if self.point_hits_vertical_bar(viewport, transformed_pos) {
                self.scroll_bars.hovered = BarHoveredState::VerticalHovered;
                self.reset_scrollbar_fade(ctx);
                true
            } else if self.point_hits_horizontal_bar(viewport, transformed_pos) {
                self.scroll_bars.hovered = BarHoveredState::HorizontalHovered;
                self.reset_scrollbar_fade(ctx);
                true
            } else {
                if self.scroll_bars.hovered.is_hovered() {
                    self.scroll_bars.hovered = BarHoveredState::None;
                    self.reset_scrollbar_fade(ctx);
                    // we need to invalidate here manually, because we're returning
                    // false (continuing to propogate this event) _and_ we need to be drawn.
                    ctx.invalidate();
                }
                false
            }
        }
        Event::MouseDown(event) => {
            let transformed_pos = event.pos + self.scroll_offset;
            if self.point_hits_vertical_bar(viewport, transformed_pos) {
                self.scroll_bars.held = BarHeldState::VerticalHeld(
                transformed_pos.y - self.calc_vertical_bar_bounds(&viewport).y0,
                );
                true
            } else if self.point_hits_horizontal_bar(viewport, transformed_pos) {
                self.scroll_bars.held = BarHeldState::HorizontalHeld(
                transformed_pos.x - self.calc_horizontal_bar_bounds(&viewport).x0,
                );
                true
            } else {
                false
            }
        }
        _ => false,
    };
    
    if is_handled {
        ctx.invalidate();
        ctx.is_handled();
        return;
    }
    
    // send the event down to the child
    let child_event = event.transform_scroll(self.scroll_offset, viewport);
    if let Some(child_event) = child_event {
        self.child.event(&child_event, ctx, data, env)
    };
    
    // handle other events after child, as appropriate
    
    match event {
        // Show the scrollbars any time our size changes
        Event::Size(_) => self.reset_scrollbar_fade(ctx),
        // The scroll bars will fade immediately if there's some other widget requesting animation.
        // Guard by the timer id being invalid.
        Event::AnimFrame(interval) if self.scroll_bars.timer_id == TimerToken::INVALID => {
            // Animate scroll bars opacity
            let diff = 2.0 * (*interval as f64) * 1e-9;
            self.scroll_bars.opacity -= diff;
            if self.scroll_bars.opacity > 0.0 {
                ctx.request_anim_frame();
            }
        }
        Event::Timer(id) if *id == self.scroll_bars.timer_id => {
            // Schedule scroll bars animation
            ctx.request_anim_frame();
            self.scroll_bars.timer_id = TimerToken::INVALID;
        }
        // we give the child a chance to handle scroll-wheel events first
        Event::Wheel(wheel) if !ctx.is_handled() => {
            if self.scroll(wheel.delta, size) {
                ctx.invalidate();
                ctx.set_handled();
                self.reset_scrollbar_fade(ctx);
            }
        }
        _ => (),
    }
}
```

So... i dunno. This is definitely a big ole' function, in any case.

Reply:
:man_shrugging: